스터디그룹 서브탭 작업

Na2te opened the pull request:
## 📌 연결된 이슈
- #86 

### ✅ 작업 내용
- 서브 탭 화면
- 공지 생성 화면
- 투표 생성 화면
- 선택된 탭에 따라 생성 화면 분기처리
- 투표 리스트 API 연동 + Paging3
- 공지 리스트 API 연동 + Paging3
- 투표 상세 화면 관련 불러오기, 참여 등 API 및 기타 로직들 처리
- 공지 상세 화면 관련 API 및 기타 로직들 처리

### 📷 관련 이미지

|영상|투표 종료시|
|:--:|:--:|
|<video src="https://github.com/user-attachments/assets/7a754f30-a916-4c5f-90ca-3ad06cb02381" alt="시연" controls></video>|<img width="250" alt="image" src = "https://github.com/user-attachments/assets/15bd2fdc-9cc5-480f-a5cf-7fde09506a54" />|






### 🤔 의논하고 싶은 내용(코드 리뷰 받고 싶은 부분)
- **theme의 NeegongNaeGongColors~가 detekt에 걸려서 Color로 바꿨습니다     https://github.com/NeeGongNaeGong/android/commit/cee2ff9c63a2314526bc742b1a9ffad89bb7545d 커밋을 참고해주세용**
- 브랜치 이름을 보시면 알겠지만 투표 리스트만 api만 하려 했지만, 관련해서 연관된 작업들도 많고, 그에 비해 기획된 건 없지만, 저것만 했다간 너무 지체될 거 같아서 그냥 하는김에 하다보니 없는 부분은 상상 기획 및 디자인과, 무아지경으로 코딩을 해버렸습니당..
- **위의 이유로 인해 솔직히 정상적인 코드 리뷰가 어려울 거라고 인지하고 있습니다. 코드 리뷰는 하지 않으시더라도, 위의 Color 관련 변경사항만 꼭 인지해주시면 감사하겠습니다**
- 솔직히 코드가 깔끔하지 않습니다, 정신없이 구현하느라 아쉬운 부분들이 많을 것으로 예상되는데 혹시나 코드 리뷰를 진행하신다면 분야를 막론하고 아쉬운 점이 보이신다면 망설이지 말고 커멘트 날려주시면, MVP 이후든 리팩토링 이슈로 파서 진행하겠습니다
- 아직 공지, 투표 수정 및 삭제 관련 내용이 논의되지가 않아서 그 부분은 논의되면 추가로 구현하겠습니다

Na2te committed:
feat(#86): 화면 경로 추가
- 투표 화면과 공지사항을 보는 리스트 화면 경로 추가

Na2te committed:
feat(#83): ListContract 생성
- List에서 현재 몇번째 인덱스(선택된 탭 인덱스)를 가리키고 있는지를 나타내는 상태 등 정의

Na2te committed:
feat(#86): ListViewModel 생성
- Contract의 State를 초기화 하는 코드 작성

Na2te committed:
feat(#86): 리스트 내 Tab 생성
- 공지와 투표 탭 생성

Na2te committed:
feat(#86): Contract 생성
- ListContract 생성
- TabIndex를 구분하기 위한 EnumClass 생성
- Tab 클릭 이벤트 처리 위한 Event와, 글 생성 이벤트 처리 위한 이벤트 생성

Na2te committed:
feat(#86): ViewModel 설정
- handleEvent 처리 및 state 초기화 수정

Na2te committed:
feat(#86): List 화면 수정
- Tab을 EnumClass로 구현해서 기존 Immutable 리스트 제거
- Tab 수정했을 때 Click Listener 구현
- 글 생성 관련 Click Listener 구현

Na2te committed:
feat(#86): LocalDateTimeAdapter 수정
- deserialize 하는 과정에서 밀리초까지 들어오는 것도 변환할 수 있도록 수정

Na2te committed:
feat(#86): 투표 리스트 API 추가
- 각 스터디그룹의 투표 리스트를 가져오는 API 추가

Na2te committed:
feat(#86): NetworkStudyGroupDataSourceImpl 수정
- API를 통해 스터디별 투표 목록을 가져오는 함수 추가

Na2te committed:
refactor(#86): Cursor 키 수정
- Cursor Key data class 이름 수정됨에 따라 MemberStudyContentsPagingSource 내의 키 이름 변경
- CursorSliceKey로 Cursor dataclass 이름 변경

Na2te committed:
feat(#86): 데이터 정의
- 스터디 그룹의 투표 목록의 요청 및 결과를 받기 위한 data class 생성

Na2te committed:
feat(#86): Paging3 관련 데이터 추가
- API 호출시 반환값인 StudyGroupVoteListBySliceResponse 생성
- Paging 관련 로직이 들어가 있는 StudyGroupVoteListPagingSource 구현

Na2te committed:
feat(#86): StudyGroupRepository 수정
- 투표 목록 가져오는 함수 추가

Na2te committed:
feat(#86): StudyGroupVoteListInfo 추가
- 각 그룹의 투표에 대한 정보를 가져오는 StudyGroupVoteListInfo 생성

Na2te committed:
feat(#86): GetVoteListUseCase 생성
- 스터디 그룹의 투표 목록을 가져오는 UseCase 생성

Na2te committed:
feat(#86): StudyGroupList에서 사용할 Paging3 상태 화면 생성
- 로딩 상태, 에러 상태 등에 따라 보여줄 화면 작성

Na2te committed:
feat(#86): VoteCard 생성
- 투표 목록들을 보여주기 위한 각 투표를 보여줄 Card 화면 구현

Na2te committed:
feat(#86): 투표 목록 화면 및 로직 구현
- 투표 목록 리스트로 보여주는 화면 및 기타 로직 구현

Na2te committed:
feat(#86): StudyGroupListGraph 생성
- 공지와 투표를 보여주는 스터디 그룹 내의 탭 내비게이션을 설정한 확장함수 구현

Na2te committed:
feat(#86): AppNavigation 경로 추가
- Studies의 공지, 투표들을 보여주는 List 생성
- 투표 탭과 공지 탭 등을 대표하는 Notice, Vote 생성
- 각 탭에서 목록을 보여주는 NoticeList, VoteList 생성
- 구체적인 내용을 보여주는 NoticeDetail, VoteDetail 생성

Na2te committed:
feat(#86): StudiesNavGraph 경로 추가
- 공지사항과, 투표 등을 보여주는 리스트 경로 추가

Na2te committed:
feat(#86): List 구체적인 로직 구현
- 이벤트 등 구체적인 로직 구현

Na2te committed:
fix(#86): Studies.List.Main 수정
- title 속성 추가

Na2te committed:
fix(#86): StudiesNavGraph 수정
- List.Main으로 들어갈 때 타이틀 인자를 받아서 List의 TopAppBar의 title로 사용할 수 있도록 설정

Na2te committed:
fix(#86): 리스트 수정
- title을 받아서 TopAppBar의 타이틀로 설정

Na2te committed:
fix(#86): 투표 목록 조회 API 데이터 타입 수정
- LocalDate 형식이던 endTime의 형식을 LocalDateTime으로 변경

Na2te committed:
fix(#86): 카드 화면 수정
- 종료시간과, 자신의 투표 유무를 인자로 받아 투표 목록의 상태에 반영

Na2te committed:
fix(#86): 데이터 타입 수정
- 투표 목록의 id를 가져와서 click 시 id 관련 작업을 할 수 있도록 수정( 투표 디테일 접근 )

Na2te committed:
feat(#86): 공지 목록 조회 api 생성

Na2te committed:
feat(#86): 공지 조회 datasource 함수 추가
- 공지 조회 api를 실행하는 함수 datasource 내부에 구현

Na2te committed:
feat(#86): 공지 목록 모델 추가 및 수정
- StudyGroupNoticeListInfo 생성 및 적용
- NetworkStudyGroupDataSource에서 공지 목록 호출 인자로 StudyGroupNoticeListInfo 사용
- 공지 목록 Slice Response를 받을 StudyGroupNoticeListBySliceResponse 생성

Na2te committed:
feat(#86): StudyGroupNoticeListBySliceResponse 수정
- content의 데이터를 StudyGroupNoticeHistoryResponse로 변경

Na2te committed:
feat(#86): NoticeHistoryInfo 모델 생성 및 Mapper 생성
- DataLayer에서 받은 공지 목록 리스트를 사용하는 Doamin 모델 생성
- DataLayer의 모델을 NoticeHistoryInfo로 변환하는 Mapper 생성

Na2te committed:
feat(#86): 공지 리스트 Paging 작업
- StudyGroupNoticeListPagingSource 구현
- Paging 데이터가 있는 Flow를 넘기는 getNoticeList 함수 repository 내 생성

Na2te committed:
feat(#86): LocalDateTime Converter 형식 추가
- yyyy-MM-dd HH:mm:ss.SSSSSS 형식의 LocalDateTime도 변환 가능하도록 formatter 추가

Na2te committed:
feat(#86): StudyGroupNoticeHistoryResponse 생성
- 공지사항 리스트 API 호출 결과를 받는 DataLayer 모델 생성

Na2te committed:
feat(#86): NoticeCard 생성
- 각 공지사항들을 보여줄 카드 생성

Na2te committed:
feat(#86): GetNoticeListUseCase 생성
- 공지사항 목록을 가져오는 GetNoticeListUseCase 생성

Na2te committed:
feat(#86): 공지사항 화면 및 로직 작업
- 공지사항 목록 보이는 리스트 화면 및 관련 Contract, ViewModel 생성

Na2te committed:
feat(#86): VoteCard 수정
- TextOverFlow 관련 속성 추가
- Q를 Card의 가운데에서 최상단에 위치하도록 수정

Na2te committed:
feat(#86): NoDataItem 수정
- NoDataItem인 경우가 공지인지, 투표인지에 따라서 메세지를 다르게 수정
- 이를 위해 경로로 From Enum Class를 받고, 내부의 route를 NoDataItem 메세지에 사용

Na2te committed:
feat(#86): 공지사항 리스트 경로 설정

Na2te committed:
refactor(#86): 공지, 투표 리스트 파일 이름 변경
- ~~Screen 꼴로 변경

Na2te committed:
refactor(#86): Contract, ViewModel 이름 변경
- 공지 및 투표의 리스트를 보여주던 것 관련 Contract 및 ViewModel을 ~~List~~로 변경

Na2te committed:
refactor(#86): import 경로 수정
- 파일 리네이밍하면서 바뀐 경로 수정

Na2te committed:
refactor(#86): Writer 클래스 이전
- Writer가 하나의 API가 아니라 여러 API에서 등장해서 별도의 파일로 뺌

Na2te committed:
feat(#86): 공지 디테일 API 생성 및 Usecase 생성
- 특정 공지사항의 자세한 정보를 가져오는 API 및 관련 DataSource, Repository 작업
- Mapper 생성
- Usecase 구현 완료

Na2te committed:
refactor(#86): List 리네임
- 공지와 투표 사항을 보던 스크린 이름을 List에서 ListScreen으로 변경

Na2te committed:
fix(#86): ListScreen 화면 변경
- 기존에 NavHost로 처리하던 것을 HorizontalPager로 변경
- 관련 Effect 등 불필요한 것 제거
- VoteListScreen, NoticeListScreen 불필요한 인자 제거 및 수정

Na2te committed:
fix(#86): 불필요한 경로 삭제
- 불필요한 네비게이션 경로 삭제

Na2te committed:
fix(#86): Index 위치 수정 및 불필요한 그래프 삭제
- 과거 NavHost를 위해 사용하던 불필요한 그래프 삭제

Na2te committed:
fix(#86): List의 Main 기본 값 삭제

Na2te committed:
feat(#86): Notice 상세 화면 및 로직 구현
- 공지 상세 화면 및 로직 구현

Na2te committed:
fix(#86): 네비게이션 경로 수정
- detail 경로 관련해서 인자 수정

Na2te committed:
fix(#86): 공지 디테일 모델 수정
- imgUrl이 nullable하게 들어와서 수정

Na2te committed:
fix(#86): NoticeCard 클릭 함수 추가
- 눌렀을 때 실행시킬 함수 관련 작업

Na2te committed:
feat(#86): 네비게이션 로직 추가
- List에서 Detail로 넘어가는 ClickEvent 등 관련 로직 추가

Na2te committed:
feat(#86): NoticeList 클릭 로직 추가
- NoticeList 내의 아이템 눌렀을 때 디테일로 가는 로직 관련 작업

Na2te committed:
feat(#86): 투표 디테일 api 추가
- api연동 및 관련 datasource, repository, usecase, mapper 작업

Na2te committed:
feat(#86): 투표 생성 화면 연결
- 투표 리스트 보이는 화면에서 추가 눌렀을 때 이동하도록 구현

Na2te committed:
feat(#86): 공지 생성 화면 구현

Na2te committed:
feat(#86): 공지 생성 API 구현

Na2te committed:
feat(#86): 공지 생성 API 화면 연동

Na2te committed:
fix(#86): Pager 패딩 수정
- 투표, 공지 보이는 부분 여백 추가

Na2te committed:
refactor(#86): 불필요한 인자 제거 및 생성 관련 로직 수정
- 불필요한 title 인자 제거
- 생성시 Main으로 다시 돌아가도록 설정

Na2te committed:
refactor(#86): 투표 생성시 studyGroupId 타입 수정
- Int에서 Long으로 수정

Na2te committed:
feat(#86): StudyGroupApi 수정
- 투표 상세 API 경로 오타 수정
- 투표 참여 API 생성

Na2te committed:
feat(#86): NetworkStudyGroupDataSource 수정
- 투표 참여 API 관련 로직 추가

Na2te committed:
feat(#86): StudyGroupRepository 수정
- 투표 참여 관련 함수 작성

Na2te committed:
feat(#86): StudyGroupVoteDetailInfoMapper 수정
- 투표 항목별 참여한 사람 데이터를 Domain 모델로 변환하는 Mapper 구현

Na2te committed:
fix(#86): 투표 생성 인자 변경
- multiple 등 속성들에 prefix로 is를 붙이는 걸로 수정

Na2te committed:
fix(#86): 투표 상세정보 response 수정
- 투표 생성자의 이미지 url등을 수신받도록 수정

Na2te committed:
fix(#86): StudyGroupVoteStatus 수정
- 각 옵션별 투표 ID와 투표한 사람들의 정보를 받도록 수정

Na2te committed:
fix(#86): NoticeCard, VoteCard 수정
- NoticeCard: modifier 인자 추가
- VoteCard: 클릭 시 투표 옵션의 아이디 인자 던져서 디테일로 넘어가도록 수정

Na2te committed:
style(#86): import 수정
- Contract까지 import로 넣어서 Effect.~로 사용할 수 있게 수정

Na2te committed:
feat(#86): VoteListScreen, NoticeListScreen 수정
- NoticeListScreen : fillMaxWidth 추가
- VoteListScreen : 투표 목록 눌렀을 때 관련 onClick 함수 추가

Na2te committed:
feat(#86): AppNavigation에 경로 추가
- 투표한 사람들 목록을 보는 VotedPerson 경로 추가
- 기존의 List로 명명되어있던 경로를 List 컬렉션이랑 이름이 겹치는 것 방지 위해 SubTab으로 변경

Na2te committed:
feat(#86): StudiesNavGraph 수정
- NoticeRoute, VoteRoute에서 popUp 관련 경로 및 inclusive 값 정정
- 기존의 List를 SubTab으로 경로 변경하면서 생긴 변경
- 투표한 사람들을 보여주는 화면인 VotedPerson 관련 경로 생성
- VoteDetail에서 VotedPerson으로 넘어가는 네비게이션 추가하거나

Na2te committed:
feat(#86): VoteItemsRequest 생성
- 투표 참여시 투표 목록을 리스트로 가지는 request 객체 생성

Na2te committed:
feat(#86): 투표 참여 CastVoteUseCase 생성

Na2te committed:
feat(#86): 투표 OptionButton 생성
- 각 투표 항목들을 보여주는 OptionButton 생성

Na2te committed:
feat(#86): VoteDetail 구현
- 각 투표에 대한 참여 등을 할 수 있는 VoteDetail 구현

Na2te committed:
feat(#86): VotePersonListScreen 구현
- 어떤 항목에 누가 투표했는지 볼 수 있는 화면 구현

Na2te committed:
feat(#86): Color 수정
- 기존 Color 파일의 neegongnaegongcolors로 되어 있던 것이 detekt에 막혀 Color로 수정
- 투표시 버튼 활성화 색깔인 yellow 추가

Na2te committed:
feat(#86): 선택항목 추가 API 연결

Na2te committed:
feat(#86): 선택항목 추가 관련 로직 및 화면 구현
- 선택 항목 추가가 가능한 투표에 대해서만 해당 작업 진행할 수 있게함

Na2te committed:
feat(#86): 투표, 공지 생성 로직 구체화
- 투표는 제목과, 항목이 하나라도 있을 경우
- 공지는 제목과, 컨텐츠의 내용이 있을 경우 활성화되도록 수정
- 특정 공지의 LocalDateTime의 밀리초가 5자리로 와서 에러나는 것 수정

Na2te committed:
feat(#86): LocalDateTimeAdapter 수정
- 밀리초가 3자리로 고정되어, 관련 adapter 코드 추가

Na2te committed:
feat(#86): 투표 종료 관련 로직 추가
- 종료된 투표일 때는 미참여 여부와 상관없이 투표 결과가 보임
- 투표 참여 버튼이 Disable되며 참여할 수 없음

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/model/studygroup/response/StudyGroupNoticeDetailResponse.kt:
```diff
@@ -0,0 +1,13 @@
+    val content: String,
+    val createdAt: LocalDateTime,
+    val modifiedAt: LocalDateTime,
+    val userProfileImage: String?,
```
저희 정책상 사용자 이미지 not null로 알고 있는데 (없으면 서버에서 기본 이미지) 
여기서는 서버 측에서 nullable로 전달하나요?

Na2te replied:
저도 기억력이 좋지 않은편이라 정확하게 기억이 안나는데,
제가 nullable하게 설정하는 편이 아닌데 있는거 보면,
전에 nullable하게 들어와서 터져서 저렇게 했던 거 같습니다

확실하게 nullable하지 않은게 보장됐다면 수정해야겠네용

Na2te replied:
수정했습니다
de8cfc43658b773d495cf480b675e511d9219041 커밋 참고해주시면 감사하겠습니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/datasource/network/NetworkStudyGroupDataSourceImpl.kt:
```diff
@@ -2,29 +2,95 @@ package com.ssafy.neegongnaegong.data.datasource.network
+                cursorId = request.cursorId,
+                size = request.size,
+            )
+
```
이 두 함수는 apiFlow를 사용하지 않은 이유가 있나요?

repository에서 first()로 호출하면 단발성도 가능한데 그러지 않은 이유가 궁금합니당

Na2te replied:
아 `first`라는 함수도 있네요 처음 봤습니당

근데 사실 이거 적용하는거랑 별개로 좀 아직 마음 속에 의문이라면, 사실 저것들 전부가(저것 외 기타 api 등) 전부 소켓 통신이 아닌 `Rest API` 통신으로 단발성인 것 뿐인데 `flow`로 반환할 이유를 조금 가지고 있었긴 했습니다.
`apiFlow`를 통해 `cold stream`의 `flow`를 만들기 때문에 `collect` 후 알아서 `close` 된다는 건 알고 있지만, 얻을 수 있는 장점이 무엇인지 아직 와닿지 못해서 그런 것 같습니다
`Paging3` 같은 `Hot StreamFlow`가 아니면 제가 모르는 다른 장점이 있을까요..?
전에 얘기가 나왔었는데 제가 까먹은 것 같기도요..?
(`apiFlow`를 사용하지 않은 건 단순히 `first`라는 함수 존재를 몰라서이긴 합니다)

Aloe-droid replied:
이것도 좋은 주제라고 생각합니다 
flow의 장점은 function chaining 방식을 이용한 핸들링이 쉽다는 장점을 가집니다. 
ex) `.catch` , `.map` , `retryWhen` ... 
 이를 viewModel이나 useCase에서 핸들링을 통일성있게 수행할 수 있다는건 분명한 장점이라고 생각합니다 

두번째 장점은 유지보수성입니다. 
repository에서 정책이 변해서 서버가 아닌 local db에서 데이터를 가져온다고 가정해보겠습니다 
remote에서 cold stream으로 제공한 flow를 이어받거나, local에서 hot stream으로 제공한 flow를 이어받아도 repository에서는 항상 Flow<T>를 반환하기에 repository에서의 수정을 최소화할 수 있습니다. 
다만 이 경우는 수정하기 어려운 과정이 아니라 큰 필요성을 느끼진 않았습니다. (저는 1번의 이유가 크다고 생각합니다.)

단점은 말씀해주신 부분처럼 **단발성 작업이 자명**한데 굳이 flow로 반환할 필요가 있을까 입니다. 
그 경우 위 예시 처럼 `.first` , `.single` , `.last` 로 단발성으로 강제해야 하는데 그럴꺼면 애초에 `suspend`로 수행하면 되지 `flow`가 필요 없지 않나라고 생각할 수 있습니다. 

두 방법 모두 장단점이 존재하는데 니공내공의 프로젝트는 기본이 `flow`인 것으로 알고 있어서 팀 규약에 맞게 수행하면 좋지 않을까 생각합니다.

Na2te replied:
functioning chaning 보니까 저번에 얘기해주셨던 게 이제 떠올랐습니다
덕분에 말고도 다른 내용도 알 수 있었네용 좋은 정보 감사합니다!

수정했습니다
f39ef6b6ff3b1f358be8e7725bc7d9c6f86cf3e4 커밋 참고해주시면 감사하겠습니다!

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/remote/adapter/json/LocalDateTimeAdapter.kt:
```diff
@@ -1,26 +1,45 @@
+            }
+        }
+        throw JsonParseException("Unparseable date: ${json.asString}")
     }
```
저희 서버측과 이야기해서 마지막 세자리 `SSS`까지로 알고있는데 이 부분 한번 더 확인해주시면 감사하겠습니다.
@orangeshop

Na2te replied:
기존에 서버에서 각 `API` 응답 결과마다 `LocalDateTIme`의 들어오는 형식이 다 달랐습니다
그래서 그걸 전부 대응하기 위해서 그때 그때 필요한 `formatter` 종류들을 넣다보니까 저렇게 되긴 했습니다

오늘 점심 즈음에 3자리로 바뀌었다는 얘기를 들었습니다
다만 3자리로 바뀌었다고 했지만 테스트 하던 도중 아직 안 바뀐 값도 있는 것을 확인했었습니다 (지금은 수정됐습니다)
아예 확실히 안쓰는게 확인되면 저렇게 list로 관리할 필요는 없긴 한데, 위와 같은 경우가 있을 수 있을 것 같아서 혹시 몰라서 안지우고 넣어놨습니다

지우는게 나을 것 같으면 지우겠습니다!

Aloe-droid replied:
저는.. 개인적으로는 서버와 규약이 정해져있고 그 부분에 맞게 오지 않는다고 클라이언트에서 에러 핸들링으로 처리하게 되면 추후  잘못된 상황이 발생했을때 에러 찾기가 힘들어지지 않을까 생각합니다. 
ex) db에 3자리 5자리 여러 자리로 저장되어서 다른 곳에서 에러가 나는 경우 

그치만 그런 상황이 있을지...는 잘 모르겠습니다 ㅋㅋㅋㅋ..... 
다른 분들의 생각을 들어보셔도 좋구 그냥 편하신대로 수정안하셔도 괜찮다고 생각함니당

Na2te replied:
@upsk1 @todayis-sunny 코드 리뷰 안하셔도 됩니다만 혹시 이부분에 대해서 어떻게 생각하실까요..?

upsk1 replied:
지금 생각해봤을때 자리수가 다르고, 그걸 모른채로 작업한다고 나중에 버그가 터졌을때 못찾을 상황은 없을걸로 생각합니다. 쉽게 말하면 자리수 다른부분 추적이 가능할 것 같아요 
그에 비해서 처리를 안해놨을때, 제가 직접 앱이 뻗는걸 경험했습니다. 그래서 최소한의 안전장치로 남겨두는걸 지향합니다요

upsk1 replied:
근데 제 짧은 견해니까 승준님 말대로 못찾을 상황이 있을 것 같아요, 하지만 짧은 견해대로 말씀드려봤습니다

todayis-sunny replied:
> 그에 비해서 처리를 안해놨을때, 제가 직접 앱이 뻗는걸 경험했습니다. 그래서 최소한의 안전장치로 남겨두는걸 지향합니다요

저도 동의합니다!

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/repository/StudyGroupRepositoryImpl.kt:
```diff
@@ -4,35 +4,83 @@ import androidx.paging.Pager
+    @Inject
+    constructor(
+        private val dataSource: NetworkStudyGroupDataSource,
+    ) : StudyGroupRepository {
```
저희 repository layer에서 dispatcher를 IO 로 변환하는 것으로 통일한것으로 알고있습니다.
만약 그렇지 않다면 Dispatchers.main 에서 코루틴 함수들이 수행될텐데 확인해주시면 감사하겠습니다.

대부분 viewModel Scope에서 launch 하실 때 IO로 변환하셨네용
그래도 통일성을 생각해보았을때 repository에서 io로 변환하는 것도 좋다고 생각함니다

Na2te replied:
아 repository에서 통일한지 몰랐습니다
확인하신 것처럼 코루틴 빌더 호출 시에 IO로 할당했습니다
이건 수정해놓겠습니다!

Na2te replied:
수정했습니다
72e03bbf31e56efd9d2c3fc4e3273e393a1e791f 커밋 참고해주시면 감사하겠습니다!

Na2te committed:
Merge branch 'develop' into feat/#86-votelist

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/data/datasource/network/NetworkStudiesDataSource.kt:
```diff
@@ -1,9 +1,17 @@
+    suspend fun createVote(
+        studyGroupId: Long,
+        requestBody: CreateVoteRequest,
+    ): Flow<Unit>
```
혹시 여기서 Flow를 쓰는데 suspend를 동반한 이유가 있으신가요 ??

Na2te replied:
아뇨 불필요합니다
왜 그랬는지 모르겠네요 수정했습니다
136549010e99f6fe3b5e80b197ababb2fd22b547 커밋 참고해주시면 됩니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/studygroup/StudyGroupVoteStatusInfo.kt:
```diff
@@ -0,0 +1,14 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class StudyGroupVoteStatusInfo(
```
만약 `@Serializable`의 이유가 화면에서 bundle로 전달하기 위함이라면 저는 UiModel을 따로 정의하는 게 좋다고 생각합니다.
화면간 전달에서 사용되는 데이터라면 Compose에서 `@Stable` 어노테이션이 필요할 수 있는데 (skip composition을 위해) 그런 부분을 대응하려면 UiModel <-> Domain Entity를 분리하는게 좋다고 생각합니다. 

물론 화면 전환간 데이터 전달을 위해서가 아니라면 무시하셔도 죠습니다!!

Na2te replied:
말씀하신 목적대로 본 데이터를 Navigate시 넘겨야 해서 썼었습니다만... 해당 객체의 리스트를 던져야 하는데 저걸로는 해결이 되지 않더군요
그래서 Json으로 String꼴로 변환한 내용을 던졌다가 다시 복구하는 식으로 구현하면서 Serializable은 필요가 없어졌습니다
일단 어노테이션은 지워야겠네용

그것과 별개로 Stable한 것 대응이라면,
본 데이터는 Navigation으로 이동할 때 인자로써 전달되는 것이며, 이걸 인자로 받는 화면에서는 어떠한 상태도 가지고 있지 않습니다 (ViewModel 등이 없고 그저 받은 데이터를 화면에 보여주는게 전부)
그래서 Recomposition이 일어날 일이 없는 화면인데, 그럼에도 UiModel을 생성하는게 좋을까요?

Aloe-droid replied:
1. 인자로써만 활용되고 화면에서 상태를 가지지 않는다면 문제가 생기지 않습니당 
2. 화면 전달 간에 사용되는 데이터가 UI에 보여질 수 있기에 저는 `UiModel`을 분리해서 사용하는 편입니다. (혹시 모르니까)
근데 해당 부분이 규약으로 정해진 것도 없고, 리컴포지션에 영향을 끼치지 않는다면 수정하지 않아도 괜찮다고 생각합니다.

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/studygroup/StudyGroupVoteListInfo.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class StudyGroupVoteListInfo(
+    val studyGroupId: Long,
+    val cursorTime: LocalDateTime?,
```
cursorTime이 Null로 올때가 있나요 ??

Na2te replied:
제가 `DataClass` 이름을 조금 잘못 지은 것 같습니다 `Info`가 아니라 `Request`로 명명하는게 맞는 거 같은데 잘못했네요
해당 클래스 자체는 투표 목록을 슬라이싱으로 가져오기 위한 것인데, 
**최초 or refresh** 되어, 쓸 수 있는 Cursor 값이 없을 때 null로 보내주어 최초 커서 데이터를 받게 됩니다.

이름을 잘못 지은 감이 있어서 `~Info`를 `~Request`로 rename하겠습니다 감사합니다!

Na2te replied:
현재 수정했습니다
해당 파일 외 다른 ~Info로 되어 있던 Cursor 관련 Request 모델 이름 또한 같이 수정 완료했습니다
6dd77822ff308c0b3c98733b935c42218291f99a 참고해주시면 될 것 같습니다

upsk1 replied:
아하 이해했습니다

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/studygroup/StudyGroupVoteStatusInfo.kt:
```diff
@@ -0,0 +1,14 @@
+) {
+    @Serializable
+    data class VotedMemberInfo(val id: Long, val username: String, val profileImg: String)
+}
```
![image](https://github.com/user-attachments/assets/688bf974-7bba-4eae-bc4e-7cc07f894593)

이런 장점이 있네요 배우고갑니다

Aloe-droid replied:
저도 처음봤는데 신기하네요 배우고 갑니다 ..

Na2te committed:
refactor(#86): NetworkStudiesDataSource 수정
- 불필요한 suspend 키워드 제거

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/component/Paging3Component.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+enum class From(val route: String) {
+    Notice("공지"),
+    Vote("투표"),
+}
```
한국어로 강제하는 건 편하지만 다른 언어로 변환하기 힘들다고생각합니다 .
res -> strings 에서 관리하고 이곳에서는 id 값(int)만 전달하고 화면에서 `stringResource`로 호출하는건 어떨까용?

다른 코드의 대부분이 단순 문자열로 되어있네여.. 이건 나중에 몰아서 바꾸눈게 좋을수도 있다고 생각합니당..

Na2te replied:
저희 앱이 다국어 지원까지 고려하고 있는지를 모르겠습니다
이미 TopAppBar등 대부분이 하드코딩한 텍스트를 쓰고 있는 거로 알고 있고, 다국어 지원 생각 안하는 거로 알고 있어서 저렇게 작성했습니다
다국어 지원 고려하게 되면 MVP 이후 리팩토링 해야 되지 않을까 생각하고 있었습니당

Aloe-droid replied:
아하  죠습니다

todayis-sunny replied:
저도 `stringResource`랑 직접 하드코딩 **String**이 섞여 있습니다🫣

Na2te committed:
refactor(#86): Cursor Slicing Request 모델 이름 정정
- ~Info로 되어 마지 response 모델처럼 되어 있던 이름을 ~request로 수정

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/main/ListScreen.kt:
```diff
@@ -0,0 +1,218 @@
+                    is Effect.NavigateToMakeVote -> navigateToMakeVote()
+                }
+            }
+        }
```
Side Effect를 CollectLatest 하는것이 위험할 수 있다고 생각합니다 .

UiState는 항상 최신 상태를 유지해야 하는 만큼 Latest가 일반적이지만 
Effect는 로직이 수행되는 부분이라 아래와 같은 예시가 생길 수 있습니다 .
`Effect A 작업 수행 -> Effect B 작업 방출 -> A 작업 중단 -> B 작업 수행`
A 작업이 중요하다면 강제로 종료되는 만큼 위험할 수 있어서 제 생각에는 SideEffect는 Collect가 좋다고 생각합니다. 
이부분은 다른 분들의 의견도 궁금하네요

Na2te replied:
이 부분은 크게 생각이 없었습니다
수정하고 오겠습니당

Na2te replied:
수정했습니다!
d54e5847985da87d6a4f5c4c46899a124ae51214 커밋 참고해주시면 감사하겠습니다

추가로 다른 것 수정하면서 발견한 것들이 있어서 추가로 작업했습니다
c67defb60431ffa33f4fe30a8be135367e370446, 0ef8232afb0ce2c8463c4bd962490aa4eee237e8 커밋 참고해주시면 감사하겠습니다

좀 커밋 많아져서 번거로운 거 같은데 이거 3개를 하나로 합칠까요..?

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/main/ListViewModel.kt:
```diff
@@ -0,0 +1,85 @@
+        val noticeListFlow: Flow<PagingData<NoticeHistoryInfo>>
+        val voteListFlow: Flow<PagingData<VoteHistoryInfo>>
+
+        init {
```
보통 다들 뷰모델 이닛위치 맨아래에 두나여? 전 여태 맨위에 두고 쓰는데

Na2te replied:
음 이건 큰 생각 없었습니다
init을 위에 두는게 맞을까요?
틀린건지 다른건지 모르겠지만, 이건 의견들을 듣고 싶긴 하네요
다들 희망한다면 고치면 될 것 같습니다!

Aloe-droid replied:
![image](https://github.com/user-attachments/assets/4228a24e-c67d-405b-a389-fe22fa29798c)
https://kotlinlang.org/docs/coding-conventions.html#class-layout
Kotlin 공식 컨벤션에서는 맨위라고 표기하였는데... 바꿀필요까지는 잘 모르겠습니다 
저는 편하신대로 하시면 된다고 생각함니다

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/notice/NoticeDetailScreen.kt:
```diff
@@ -0,0 +1,146 @@
+                        alignment = Alignment.Center,
+                    ),
+                requestOptions = { RequestOptions().diskCacheStrategy(DiskCacheStrategy.ALL) },
+                failure = {
+                    // 이미지 로드 실패 시 플레이스홀더
```
아니 이런게 있었군요..

todayis-sunny replied:
https://github.com/skydoves/landscapist
여기 좋은 라이브러리 가져다 쓰니 있더라구여

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/notice/NoticeListScreen.kt:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        when {
+            lazyItems.loadState.refresh is LoadState.Loading -> item { LoadingItem() }
+            lazyItems.loadState.append is LoadState.Loading -> item { LoadingItem() }
+            lazyItems.loadState.refresh is LoadState.Error ->
+                item {
+                    val e = lazyItems.loadState.refresh as LoadState.Error
+                    ErrorItem(e.error.localizedMessage.orEmpty()) { lazyItems.retry() }
+                }
+
+            lazyItems.loadState.append is LoadState.Error ->
+                item {
+                    val e = lazyItems.loadState.append as LoadState.Error
+                    ErrorItem(e.error.localizedMessage.orEmpty()) { lazyItems.retry() }
+                }
+
+            lazyItems.itemCount == 0 ->
+                item {
+                    NoDataItem(From.Notice)
+                }
```
배워갑니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/main/ListScreen.kt:
```diff
@@ -0,0 +1,218 @@
+            )
+        },
+    ) {
+        Index.entries.toImmutableList().forEach { item ->
```
toImmutableList가 있으면 추가되는게 있나요 신기하네용

Na2te replied:
음 그냥 당시에는 최대한 stable하게 만들고자 썼었던 거 같은데, 다시 돌이켜보니 필요 없는 거 같아서 그냥 없앴습니다...ㅋㅋ...
7a81db98e89180153d58a226d38982272a671ec1 커밋 참고해주시면 감사하겠습니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/notice/NoticeDetailScreen.kt:
```diff
@@ -0,0 +1,146 @@
+    content: String,
+) {
+    Column(
+        verticalArrangement = Arrangement.spacedBy(NeeGongNaeGongTheme.paddingScheme.sp3),
```
아니 paddingScheme가 있었군요... 지금까지 그냥 dp 만들어서 썻는데..... 😢

Na2te replied:
저도 최근에 알아가지고 쓰긴 했는데, 여전히 하드코딩한 부분도 있습니다
Figma 디자인부터가 이런 paddingScheme를 적용해서 썼다면 모르겠는데, 지금 작업하고 있는 부분은 디자인이 없는 부분이기도 하고 그래서 아무튼 어렵네용

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/main/ListViewModel.kt:
```diff
@@ -0,0 +1,85 @@
+        val noticeListFlow: Flow<PagingData<NoticeHistoryInfo>>
+        val voteListFlow: Flow<PagingData<VoteHistoryInfo>>
+
+        init {
+            val groupId: Long? = savedStateHandle["groupId"]
+            val startTabIdx: Int? = savedStateHandle["startTab"]
+            if (groupId != null && startTabIdx != null) {
+                setState { copy(groupId = groupId) }
+
+                noticeListFlow =
+                    getNoticeListUseCase(groupId).cachedIn(
+                        viewModelScope,
+                    )
+
+                voteListFlow =
+                    getVoteListUseCase(groupId).cachedIn(
+                        viewModelScope,
+                    )
+            } else {
+                noticeListFlow = flow {}
+                voteListFlow = flow {}
+                showErrorMessage("잘못된 접근입니다")
+                setEvent(ListContract.Event.InvalidAccess)
+            }
+        }
+    }
```
init 에서 Collect를 수행하게 되면 아래 시나리오에서 정상적으로 작동하나여 ? 

1. 공지사항 리스트 화면에서는 3개의 공지사항이 있습니다. 
2. 새로운 공지사항을 만듭니다. 
3. 공지사항 리스트 화면에 다시 오면 4개의 공지사항이 있습니다.
---
1. 공지사항 리스트 화면에서는 3개의 공지사항이 있습니다. 
2. 첫번쨰 공지사항에서 데이터들을 수정합니다.
3. 공지사항 리스트 화면에 다시 오면 1개의 변경된 공지사항과 두개의 기존 공지사항이 보입니다.

Na2te replied:
우선 생성했다가 돌아갈 때는, popUp에 inclusive로 Main 탭을 다시 로딩해서 viewModel도 다시 갱신이 됐는데(시연 영상에서도 확인 가능하실겁니당)

공지 수정은 아직 수정 관련 스크린이나 그런걸 만들지 않아서 뭐가 없는 상태입니다
수정 화면에서 수정 완료 후 Main으로 이동할 때 이 역시 inclusive로 Main을 다시 로딩하게 하면 괜찮지 않을까요..?

Aloe-droid replied:
아 탭을 지우고 다시 생성해서 `viewModel`의 `init`이 다시 호출되는거군요.
이해되었습니다. 저는 `init` 블럭에서 `collect`하게 되면 수정사항에 대응이 안될 것이라 생각해서 그랬습니다.

Na2te replied:
혹시 더 나은 방법 or 의견이 있으시다면 제시해주시면 감사하겠습니다!

Aloe-droid replied:
저는 수정 가능한 페이징 데이터를 위해서 `RemoteMediator`를 사용했었습니다.
https://www.notion.so/1e7e1a802a708023bcf2fec7e0fab4b8 
제 글입니다 ㅎㅎ.... 

다만 이미 진행하신 코드도 잘 작동될 것 같은데 나중에 리팩토링 때 한번 고민해주셔도 좋을 것 같습니다.

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/vote/VoteDetailContract.kt:
```diff
@@ -0,0 +1,70 @@
+        val progressTime: String,
+        val voteTitle: String,
+        val state: Boolean,
+        val selected: PersistentList<StudyGroupVoteDetailInfo.VoteValue>,
+        val voteOptions: PersistentList<VoteOptions>,
+        val voteItems: PersistentList<StudyGroupVoteStatusInfo>,
+        val voteValues: PersistentList<StudyGroupVoteDetailInfo.VoteValue>,
```
이런 방법으로도 Stable하게 할 수 있군여 신기합니다 ㄷㄷ

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/vote/VoteDetailViewModel.kt:
```diff
@@ -0,0 +1,172 @@
+                setEvent(Event.InvalidAccess)
+            }
+        }
+
```
저희 Route가 객체라서 navigation에서 safe args 처럼 사용할 수 있어서 아래 처럼 호출하면 not null로 받을 수 있지 않을까여?
```kotlin
val 라우트클래스: 클래스명 = savedStateHandle.toRoute()
val groupId = 라우트클래스.groupId
```

Na2te replied:
와우 이런게 있군요 감사합니다 이제 알았어요 수정하겠습니다

Na2te replied:
수정했습니다
93d59f083e96954400d516b6bee90fde81438b6a 커밋 참고해주시면 감사하겠습니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/vote/component/OptionButton.kt:
```diff
@@ -0,0 +1,193 @@
+        modifier =
+            modifier
+                .fillMaxWidth()
+                .padding(0.dp)
```
제로 패딩이 존재하는 이유가 있나용?

Na2te replied:
전에 default로 공간이 남는 것 관련해서 처음에 시도했었던 것인데 지우는 걸 깜박했습니다
수정했습니다
8b23f57eefbc8ae415b566cb68a55ac763924eef 커밋 참고해주시면 감사하겠습니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/vote/component/OptionButton.kt:
```diff
@@ -0,0 +1,193 @@
+            ) {
+                Row(modifier = Modifier.weight(1F)) {
+                    if ((!castMode || !state) && isSelected) {
+                        val iconSizeDp: Dp =
+                            with(LocalDensity.current) {
+                                NeeGongNaeGongTheme.typography
+                                    .labelMedium
+                                    .fontSize
+                                    .toDp()
+                            }
```
이렇게 하면 글자 크기에 비례해서 icon 크기가 동적으로 할당되겠네요 ㄷㄷ

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/notice/NoticeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+                    text = "공지 만들기",
+                )
+            },
+            onNavigationClick = { popBackStack() },
```
귀찮으시겠지만... popBack은 SideEffect에서만 강제하려면 이곳에서도 Event로 전달하는게 좋지 않을까 생각합니다..

Na2te replied:
아 그렇네요
지당하신 말씀입니다 
당연히 수정해야 하는 부분이네용

수정했습니다
25c36d70007d1b3bd9618e2b90f8f249a1fb2c27 커밋 참고해주시면 감사하겠습니다

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/notice/NoticeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+        )
+        HorizontalDivider()
+        Spacer(modifier = Modifier.height(NeeGongNaeGongTheme.paddingScheme.sp2))
+        BasicTextField(
+            value = content,
+            onValueChange = onChangeContent,
+            modifier =
+                Modifier
+                    .fillMaxWidth()
+                    .weight(1F)
+                    .padding(0.dp),
```
여기서 0padding 주면 TextField 패딩 이 없어지나요??

Na2te replied:
기본 TextField가 padding이 있던 것 해결하고자 넣었었는데 결론적으로 이거랑 관련 없고, BasicTextField 선언해서 해결했습니다
수정했습니다
8b23f57eefbc8ae415b566cb68a55ac763924eef 커밋 참고해주시면 감사하겠습니다

Aloe-droid reviewed (commented):
3000줄이 넘는 방대한 코드네요... 
심지어 화면에 로직이 복잡해서 고생하셨을게 눈에 보이니까 더 감사해지네요..
고생많으셨습니다

Na2te committed:
fix(#86): LocalDateTimeAdapter 중괄호 빠진 것 수정

Na2te committed:
fix(#86): 중복 import 제거
- StudyGroupApi에서 Get이 중복으로 import되어 있던 것 제거

Na2te committed:
fix(#86): StudyGroupRepositoryImpl ioDispatcher 추가
- 왜인지 모르겠지만 인자가 사라져 있어서 추가

Na2te committed:
fix(#86): userProfileImage Nullable 변경
- StudyGroupNoticeDetailInfo의 userProfileImage 속성 NonNullable하게 설정
- StudyGroupNoticeDetailResponse의 userProfileImage 속성 NonNullable하게 설정

Na2te committed:
refactor(#86): Dispatcher 위치 변경
- viewModelScope launch할 때 지정하던 Dispatcher를 repository로 변경

Na2te committed:
refactor(#86): DataSource Return 값 변경
- Paging을 위해 쓰이는 API의 return 값을 Flow로 변경

Na2te committed:
fix(#86): ListScreen side effect 수정
- collectLatest에서 collect로 수정

Na2te committed:
fix(#86): 불필요한 immutableList 변환 제거
- entries로도 forEach를 진행할 수 있고, recomposition과 관련 없어서 제거

Na2te committed:
fix(#86): ViewModel arg 가져오는 방법 수정
- toRoute를 이용해 Nullable하지 않은 인자 가져오는 방법으로 수정

Na2te committed:
fix(#86): 불필요한 padding 제거
- 0 padding 제거

Na2te committed:
fix(#86): SideEffect collect 수정
- collectLatest로 되어 있던 것 collect로 수정

Na2te committed:
fix(#86): popBackStack 관련 로직 수정
- AppBar에 실행 함수를 넘기던 것을 Event를 전달하는 것으로 수정

Na2te committed:
fix(#86): NoticeDetailScreen SideEffect collect 수정
- collectLatest로 되어 있던 것 수정

Na2te commented:
> 3000줄이 넘는 방대한 코드네요... 심지어 화면에 로직이 복잡해서 고생하셨을게 눈에 보이니까 더 감사해지네요.. 고생많으셨습니다

복잡한 건 제가 코드를 가독성 있게 짜지 못해서 그런거라고 생각합니다
더 잘 짜고 싶었지만 우선 기능 완성이 우선이라고 생각해서 구현 위주로 짰습니다
다음에 리팩토링 할 때 조금 더 가독성 있게 짜보겠습니다!

우스갯소리지만 사이즈 XL 급의 Issue 답긴 하네요ㅋㅋㅋㅋ

Na2te committed:
fix(#86): 불필요한 Serilizable 어노테이션 제거

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/data/mapper/studygroup/StudyGroupNoticeHistoryInfoMapper.kt:
```diff
@@ -0,0 +1,28 @@
+            name,
+            profileImg,
+            groupRole,
+        )
```
`StudyGroupNoticeDetailInfoMapper`에도 동일한 **mapper function**이 있는데 Mapper별로 쓰는거라면 동일하더라도 설정해주는 방향이 옳은것일까요?

Na2te replied:
어 있는 줄 몰랐습니다 음 매퍼를 StudyGroupNoticeWriterMapper로 만들까싶네용
일단 StudyGroupNoticeWriterMapper Mapper 만들어서 Writer 관련 처리 진행하고 커밋 올리겠습니다!

Na2te replied:
수정했습니다
e7ef9df62863f5baa737e30350a3869502093a83 커밋에서 관련 수정사항 확인해보실 수 있습니다 감사합니다!

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/data/remote/adapter/json/LocalDateTimeAdapter.kt:
```diff
@@ -27,13 +28,19 @@ class LocalDateTimeAdapter :
+                DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSSSS"), // 마이크로초
+                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS"), // 마이크로초
+            )
+        formatter.forEach {
+            try {
+                return LocalDateTime.parse(json.asString, it)
+            } catch (_: DateTimeParseException) {
+            }
         }
+        throw JsonParseException("Unparseable date: ${json.asString}")
```
신기하네요 `formatter` 돌면서 **Exception**이 발생하면 패스하고 다음 `formatter`로 가서 체크하는군여👍

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/notice/NoticeDetailViewModel.kt:
```diff
@@ -0,0 +1,63 @@
+        private val noticeId: Long =
+            savedStateHandle.toRoute<AppNavigation.Screen.Studies.SubTab.Screen.NoticeDetail>().noticeId
+
+        init {
+            viewModelScope.launch {
+                getNoticeDetailUseCase(groupId, noticeId).safeCollect {
+                    val outputFormatter = DateTimeFormatter.ofPattern("M월 d일 a h:mm")
+                    val createdAt = it.createdAt.format(outputFormatter)
+
+                    setState {
+                        copy(
+                            writer = it.writer.name,
+                            writerProfileImage = it.writer.profileImg,
+                            createdAt = createdAt,
+                            content = it.content,
+                        )
+                    }
+                }
+            }
```
`init`블럭을 활용하셨군여, 저는 `Evnet`에 `OnLoad`를 호출해서 **Route**단에 설정을 했는데 혹시, 둘 다 가능한 방법일까요?
뭔가 다른점은 존재할거 같긴 합니다. 관심사 분리의 차이일까요?🧐

Aloe-droid replied:
`View` 에서 `LaunchEffect`로 `load` 함수를 수행하게 되면 큰 단점으로는 잦은 서버 호출입니다.
화면이 전환되거나, 다크 모드 <-> 라이트 모드 변환 시 마다 `LaunchEffect`는 다시 실행되고 그때마다 서버에게 최신 데이터를 받아오게 됩니다. 
가장 서버와 통신이 잦게 발생하지만 그만큼 최신 데이터를 유지하는데 유리하다는 장점을 가집니다. 

`viewModel` 에서 `init` 블럭에서 데이터를 수집하게 되면 `viewModel`이 생성된 시간 이후에는 새로운 데이터를 수집하지 못하는  단점을 가집니다. 서버와의 수신 빈도가 적지만, 그만큼 최신 데이터를 가지고 있기 불리한 단점을 가집니다. 
추가적으로, 단위 테스트 시 `viewModel` 을 생성하기만 해도 서버와 통신을 강제하는 만큼 `Side Effect`가 발생할 확률을 높입니다. 

둘다 장단점이 있고 선택하기 나름이지만 아래 글에서는 google developer가 안티패턴이라고 언급했네요.. 
아래 글을 참고해주시면 감사하겠습니다.  
https://proandroiddev.com/loading-initial-data-in-launchedeffect-vs-viewmodel-f1747c20ce62

todayis-sunny replied:
오 설명 고맙습니다:)
너무 잘 이해되는것 같습니다!
> 다크 모드 <-> 라이트 모드 변환 시 마다 LaunchEffect는 다시 실행되고 그때마다 서버에게 최신 데이터를 받아오게 됩니다.

이 부분이 `스터디 Screen` -> `상세 스터디 스터디 Screen` 으로 갔을때, 다크 모드 🔄 라이트 모드 변환시 뒤로가기가 발생하는 주요 원인이 될까요?

> 둘다 장단점이 있고 선택하기 나름이지만 아래 글에서는 google developer가 안티패턴이라고 언급했네요..

읽어보니 결국엔 **cold flows for the lazy initialization**를 활용하는 방안이 좋은걸로 이해하면 될까요?
딥하게 갈수록 재미있는 만큼 어려워지긴 하네요🫠

Aloe-droid replied:
> 이 부분이 `스터디 Screen` -> `상세 스터디 스터디 Screen` 으로 갔을때, 다크 모드 🔄 라이트 모드 변환시 뒤로가기가 발생하는 주요 원인이 될까요?

이건 제가 확인해보지 않아서 잘 모르겠습니다..

> 읽어보니 결국엔 **cold flows for the lazy initialization**를 활용하는 방안이 좋은걸로 이해하면 될까요? 딥하게 갈수록 재미있는 만큼 어려워지긴 하네요🫠

최적화에 관한 이야기 인데, 적절히 서버로 부터 최신 데이터를 가져올 수 있는 방법이라 좋은 방법이라고 생각합니다. 
제가 작성한 알림, 마이페이지 스크린도 위 방식과 유사하게 작성하였습니다.

Na2te replied:
사실 데이터를 Room에 저장하고 Mediator로 가져오는게 좋긴 한데 Room 쿼리 등 작업할 게 많아져서 생략했습니다....ㅠㅠ
이래야 삼성 가는건가봐요 축하드립니다 :)
후에 리팩토링 진행해야겠네요

Aloe-droid replied:
헉.. 참말로 감사합니다 ㅎㅎ

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/list/vote/VoteDetailViewModel.kt:
```diff
@@ -0,0 +1,169 @@
+        private val addNewVoteOptionUseCase: AddNewVoteOptionUseCase,
+        private val castVoteUseCase: CastVoteUseCase,
+    ) :
+    BaseViewModel<Event, State, Effect>() {
```
```kotlin
import com.ssafy.neegongnaegong.presentation.group.list.vote.VoteDetailContract.Effect
import com.ssafy.neegongnaegong.presentation.group.list.vote.VoteDetailContract.Event
import com.ssafy.neegongnaegong.presentation.group.list.vote.VoteDetailContract.State
```
로 **import**를 하니 더 보기 깔끔해지는군요!

todayis-sunny reviewed (approved):
스터디 서브탭 작업 고생하셨습니다:)
얼른 제 작업도 PR하고 머지하겠습니다🫠

Na2te committed:
refactor(#86): 중복된 Mapper함수 제거
- Writer 클래스 관련 여러 Mapper에서 선언된 변환 함수를 별도의 Mapper에 생성
